ssl/t1_enc.c

```diff
@@ -314,6 +323,85 @@ int tls1_change_cipher_state(SSL *s, int which)
         goto err2;
     }
 
+#ifdef OPENSSL_LINUX_TLS
+    if (!(which & SSL3_CC_WRITE)) {
+#ifdef SSL_DEBUG
+        printf("\nSkipping offload for non-write context\n");
+#endif // SSL_DEBUG
+        goto skip_offload;
+    }
+    memset(&crypto_info, 0, sizeof(crypto_info));
+
+    /* check that cipher is AES_GCM_128 */
+    if (EVP_CIPHER_mode(c) != EVP_CIPH_GCM_MODE) {
+#ifdef SSL_DEBUG
+        printf("\nGot mode %08lx != GCM, skipping offload\n",
+               EVP_CIPHER_mode(c));
+#endif // SSL_DEBUG
+        goto skip_offload;
+    }
+    crypto_info.info.cipher_type = TLS_CIPHER_AES_GCM_128;
+
+    /* check version is 1.2 */
+    if (s->version != TLS1_2_VERSION) {
+#ifdef SSL_DEBUG
+        printf("\nVersion mismatch got %08x, skipping offload\n", s->version);
+#endif // SSL_DEBUG
+        goto skip_offload;
+    }
+
+    if (EVP_CIPHER_key_length(c) != TLS_CIPHER_AES_GCM_128_KEY_SIZE) {
+#ifdef SSL_DEBUG
+        printf("\nUnexpected key length %d, skipping offload\n",
+               EVP_CIPHER_key_length(c));
+#endif // SSL_DEBUG
+        goto skip_offload;
+    }
+
+    crypto_info.info.version = s->version;
+
+    /* This is not the IV that is used by OpenSSL, but rather the sequence
+     * number. We should actually provide both iv and sequence.
+     */
+    EVP_CIPHER_CTX_ctrl(dd, EVP_CTRL_GCM_GET_IV, 12, geniv);
```

Comment:
We should probably have a #define for the value 12 here...although that's not specific to this change